### DISCUSSION TITLE
Issue occuring in Azure function app but not in webapp

### DISCUSSION THREAD

**Discussion by hemanthaar** [dbosoft/YaNco#161] [Q&A] [answered]
Hi,

I am trying to use Yanco connector in my azure function app in .netcore and I am getting exception on System.Threading.Tasks.Datafow when calling BAPI in SAP. But when I use the same code as webapp, i am getting results correctly. 

Could you please help.
![issue](https://user-images.githubusercontent.com/74851652/161658502-a95af0b0-23c3-4c8e-8f36-4380a28fcbc3.jpg)

**Comment by fw2568** [comment, marked as answer]
Hi @hemanthaar , 

I converted this to discussions as it looks more like a dependency issue of Azure functions. 

YaNco uses [dbosoft.Functional](https://github.com/dbosoft/Dbosoft.Functional) for some core functional features. Dbosoft.Functional depends currently on System.Threading.Tasks.Dataflow Version 5.0. 
It looks like that another version of System.Threading.Tasks.Dataflow is used for your function app.

There are some known issues for dependencies of Azure functions apps related to System.Threading.Tasks.Dataflow. Depending on the version of Azure functions you may have to use a binding redirection (https://thewissen.io/azure-functions-binding-redirects/). But that's not a easy task. I personally prefer the new isolated azure functions as they are designed to reduce such problems. 

From the architecture perspective I would also recommend not to use any kind of SAP connector directly in a Azure function as it can scale almost endless and that not something SAP is designed for (connection limits, etc.). Please see this project for a reference architecture: https://github.com/dbosoft/SAPHub.

Hope that helps to point you in the right direction. In case you asking for more detailed information please post your Azure function version and a sample application with all dependencies used in your project. 

Please note that we also offer commercial support plans, here you can find more about it: https://support.dbosoft.eu/hc/en-us/articles/360021454080-YaNco-Support-Plans

Best Regards,
Frank

**Reply by hemanthaar** [reply]
Hi Frank,

Thanks for your valuable suggestion. 

I did import the latest functional nuget package into my application but the error still persists. We are using version 3 of Function app but the sdk version has Dataflow dll files as it says in the error. So its confusing why it is unable to load the file. Since it has the same versions, would binding redirect still be necessary?

Meanwhile I am trying with isolated functions to see if it makes any difference as you suggested.

Regards
Hemanth

![dependency](https://user-images.githubusercontent.com/74851652/162061544-cda7e98f-d876-429c-bd41-023552ee8993.jpg)
![error](https://user-images.githubusercontent.com/74851652/162061605-0a4a0a17-957b-4b23-8548-1e17586fcfd5.jpg)

**Comment by fw2568** [comment]
Hi,

I have tried to reproduce this issue with a minimal project using Azure Functions 3. You can find it on https://github.com/fw2568/YancoAzureFunctionTest/tree/master/YancoAzureFunctionTest2.

However I have no problems with any dll and SAP connection is working. 
I guess that you may have another dependency that cause a lower version of Dataflow to be selected for build. 

I can see on your screenshot that you have warning on assembly level, is this maybe related?
Please check also version of the file System.Threading.Tasks.Dataflow.dll in bin\Debug\netcoreapp3.1 (for a debug build).

It would also help if you can share the file [assemblyname].deps.json from the same directory. It contains details which version each dependency requires.

Best Reagrds,
Frank

**Reply by hemanthaar** [reply]
Hi Frank,

Thanks again.

It worked for me with latest azure function v4 this time and iIwas able to get results company code list from your examples. I am trying to use BAPI_CUSTOMER_GET_CHILDREN to get customer hierarchy data. But it gives me error on arguments cannot be inferred from the usage. I tried in different ways to resolve this but in vain. 

Can you shed some light on how to resolve. This will be of great help. Thanks.

Regards
Hemanth

![argument error](https://user-images.githubusercontent.com/74851652/162401380-fda92227-f050-45a8-a171-360de2cd9c9e.jpg)

**Reply by fw2568** [reply]
Hi,
this is caused by the position of IfLeft statement.
Move it behind the ")" for CallFunction then it will work: 

``` csharp
  var res = await context.CallFunction(
      "FUNC", f => f,
      f => f
          .HandleReturn()
          .MapTable("TAB", s =>
              from field in s.GetField<string>("FIELD")
              select new { field }

           )  // <== MapTable
           
     ) // <== CallFunction
   // returns IEnumerable with anonymous type and property field
  .IfLeft(l => throw new Exception(l.Message));
```
          
remark: please post code instead of screenshot.

**Reply by hemanthaar** [reply]
Oops!!! Thank you so much. It works perfectly now.
I really appreciate your help and being so patient with my questions.

Regards
Hemanth
